akka-stream/src/main/scala/akka/stream/scaladsl/RetryFlow.scala

```diff
@@ -0,0 +1,234 @@
+ */
+@InternalApi private[akka] final class RetryFlowCoordinator[In, Out](
+    minBackoff: FiniteDuration,
+    maxBackoff: FiniteDuration,
+    randomFactor: Double,
+    maxRetries: Int,
+    decideRetry: (In, Out) => Option[In])
+    extends GraphStage[BidiShape[In, In, Out, Out]] {
+
+  private val externalIn = Inlet[In]("RetryFlow.externalIn")
+  private val externalOut = Outlet[Out]("RetryFlow.externalOut")
+
+  private val internalOut = Outlet[In]("RetryFlow.internalOut")
+  private val internalIn = Inlet[Out]("RetryFlow.internalIn")
+
+  override val shape: BidiShape[In, In, Out, Out] =
+    BidiShape(externalIn, internalOut, internalIn, externalOut)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+
+    private var elementInProgress: OptionVal[In] = OptionVal.none
+    private var retryNo = 0
+    private var moreInternalDemand = false
+
+    setHandler(
+      externalIn,
+      new InHandler {
+        override def onPush(): Unit = {
+          val element = grab(externalIn)
+          elementInProgress = OptionVal.Some(element)
+          retryNo = 0
+          pushInternal(element)
+        }
+
+        override def onUpstreamFinish(): Unit =
+          if (elementInProgress.isEmpty) {
+            completeStage()
+          }
+      })
+
+    setHandler(
+      internalOut,
+      new OutHandler {
+
+        override def onPull(): Unit = {
+          if (elementInProgress.isEmpty) {
+            if (!hasBeenPulled(externalIn) && !isClosed(externalIn)) {
+              pull(externalIn)
+            }
+          } else moreInternalDemand = true
+        }
+
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          if (elementInProgress.isEmpty || !cause.isInstanceOf[NonFailureCancellation]) {
+            super.onDownstreamFinish(cause)
+          } else {
+            // emit elements before finishing
+            setKeepGoing(true)
+          }
+        }
+      })
+
+    setHandler(
+      internalIn,
+      new InHandler {
+        override def onPush(): Unit = {
+          val result = grab(internalIn)
+          elementInProgress match {
+            case OptionVal.None =>
+              failStage(
+                new IllegalStateException(
+                  s"inner flow emitted unexpected element $result; the flow must be one-in one-out"))
+            case OptionVal.Some((_, _)) if retryNo == maxRetries => pushExternal(result)
+            case OptionVal.Some(in) =>
+              decideRetry(in, result) match {
+                case None          => pushExternal(result)
+                case Some(element) => planRetry(element)
+              }
+          }
+        }
+      })
+
+    setHandler(externalOut, new OutHandler {
+      override def onPull(): Unit =
+        // external demand
+        if (!hasBeenPulled(internalIn)) pull(internalIn)
+    })
+
+    private def pushInternal(element: In): Unit = {
+      moreInternalDemand = false
```

Comment:
is the extra tracking needed, or could it use `isAvailable(internalOut)`?

Reply:
Yes, that works nicely.